netebpfext/net_ebpf_ext_xdp.c

```diff
@@ -59,39 +83,151 @@ ebpf_extension_data_t _net_ebpf_extension_xdp_hook_provider_data = {
 //
 
+ebpf_result_t
+net_ebpf_extension_xdp_on_client_attach(_In_ const net_ebpf_extension_hook_client_t* attaching_client)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    const ebpf_extension_data_t* client_data = net_ebpf_extension_hook_client_get_client_data(attaching_client);
+    uint32_t if_index;
+    uint32_t filter_count;
+    size_t filter_ids_buffer_size;
+    FWPM_FILTER_CONDITION condition = {0};
+    net_ebpf_extension_wfp_filter_ids_t* filter_ids = NULL;
+
+    // XDP hook clients must always provide data.
+    if (client_data == NULL) {
+        result = EBPF_INVALID_ARGUMENT;
+        goto Exit;
+    }
+
+    // validate client data.
+    if (client_data->version != EBPF_ATTACH_CLIENT_DATA_VERSION) {
+        result = EBPF_INVALID_ARGUMENT;
+        goto Exit;
+    }
+
+    if (client_data->size > 0) {
+        if ((client_data->size < sizeof(uint32_t)) || (client_data->data == NULL)) {
+            result = EBPF_INVALID_ARGUMENT;
+            goto Exit;
+        }
+        if_index = *(uint32_t*)client_data->data;
+    } else {
+        // If the client did not specify any attach parameters, we treat that as a wildcard interface index.
+        if_index = 0;
+    }
+
+    if (if_index == 0) {
+        // Client requested wildcard interface. This will only be allowed if there are no other clients attached.
+        if (net_ebpf_extension_hook_get_next_attached_client(_ebpf_xdp_hook_provider_context, NULL) != NULL) {
+            result = EBPF_ACCESS_DENIED;
+            goto Exit;
+        }
+    } else {
+        if (_net_ebpf_extension_xdp_validate_if_index(if_index) != STATUS_SUCCESS) {
+            result = EBPF_INVALID_ARGUMENT;
+            goto Exit;
+        }
+        // Ensure there are no other clients with interface index 0 or with the same interface index as the
+        // requesting client.
+        net_ebpf_extension_hook_client_t* next_client =
+            net_ebpf_extension_hook_get_next_attached_client(_ebpf_xdp_hook_provider_context, NULL);
+        while (next_client != NULL) {
+            const ebpf_extension_data_t* next_client_data = net_ebpf_extension_hook_client_get_client_data(next_client);
+            uint32_t next_client_if_index = (next_client_data->data == NULL) ? 0 : *(uint32_t*)next_client_data->data;
+            if ((next_client_if_index == 0) || (if_index == next_client_if_index)) {
+                result = EBPF_ACCESS_DENIED;
+                goto Exit;
+            }
+            next_client =
+                net_ebpf_extension_hook_get_next_attached_client(_ebpf_xdp_hook_provider_context, next_client);
+        }
+    }
+
+    // Set interface index (if non-zero) as WFP filter condition.
+    if (if_index != 0) {
+        condition.fieldKey = FWPM_CONDITION_INTERFACE_INDEX;
+        condition.matchType = FWP_MATCH_EQUAL;
+        condition.conditionValue.type = FWP_UINT32;
+        condition.conditionValue.uint32 = if_index;
+    }
+
+    // Allocate buffer to hold WFP filter Ids.
```

Comment:
```suggestion
    // Allocate buffer to hold WFP filter ids.
```
nit: consistency with line 184